runners/spark/src/main/java/org/apache/beam/runners/spark/aggregators/metrics/WithNamedAggregatorsSupport.java

```diff
@@ -0,0 +1,132 @@
+ * </p>
+ */
+public class WithNamedAggregatorsSupport extends MetricRegistry {
+
+  private MetricRegistry internalMetricRegistry;
+
+  private WithNamedAggregatorsSupport(final MetricRegistry internalMetricRegistry) {
+    this.internalMetricRegistry = internalMetricRegistry;
+  }
+
+  public static WithNamedAggregatorsSupport forRegistry(final MetricRegistry
+                                                            metricRegistry) {
+    return new WithNamedAggregatorsSupport(metricRegistry);
+  }
+
+  @Override
+  public SortedMap<String, Timer> getTimers(MetricFilter filter) {
+    return internalMetricRegistry.getTimers(filter);
+  }
+
+  @Override
+  public SortedMap<String, Meter> getMeters(MetricFilter filter) {
+    return internalMetricRegistry.getMeters(filter);
+  }
+
+  @Override
+  public SortedMap<String, Histogram> getHistograms(MetricFilter filter) {
+    return internalMetricRegistry.getHistograms(filter);
+  }
+
+  @Override
+  public SortedMap<String, Counter> getCounters(final MetricFilter filter) {
+    return internalMetricRegistry.getCounters(filter);
+  }
+
+  @Override
+  public SortedMap<String, Gauge> getGauges(final MetricFilter filter) {
+    return
+        new ImmutableSortedMap.Builder<String, Gauge>(
+            Ordering.from(String.CASE_INSENSITIVE_ORDER))
+            .putAll(internalMetricRegistry.getGauges(filter))
+            .putAll(extractGauges(internalMetricRegistry, filter))
+            .build();
+  }
+
+  private Map<String, Gauge> transformToGauges(final AggregatorMetric metric) {
+    return Maps.transformValues(metric.getNamedAggregators().renderAll(),
+        new Function<Object, Gauge>() {
+
+          @Override
+          public Gauge apply(final Object rawGaugeObj) {
+            return new Gauge<Double>() {
+
+              @Override
+              public Double getValue() {
+                // at the moment the metric's type is assumed to be
+                // compatible with Double. While far from perfect, it seems reasonable at
+                // this point in time
+                return Double.parseDouble(rawGaugeObj.toString());
+              }
+            };
+          }
+        });
+  }
+
+  private Map<String, Gauge> extractGauges(final MetricRegistry metricRegistry,
+                                           final MetricFilter filter) {
+    // find the AggregatorMetric metrics from within all currently registered metrics
+    final Optional<Map.Entry<String, Metric>> aggregatorMetricEntryOp =
+        FluentIterable
+            .from(metricRegistry.getMetrics().entrySet())
+            .firstMatch(new Predicate<Map.Entry<String, Metric>>() {
+
+              @Override
+              public boolean apply(Map.Entry<String, Metric> metricEntry) {
+                return (metricEntry.getValue() instanceof AggregatorMetric)
+                    && (filter.matches(metricEntry.getKey(), metricEntry.getValue()));
```

Comment:
Not sure if filter is necessary, but if so should filter on the internal named aggregators - the `Aggregator` name.